source/common/http/utility.h

```diff
@@ -233,6 +233,78 @@ void transformUpgradeRequestFromH2toH1(HeaderMap& headers);
  */
 void transformUpgradeResponseFromH2toH1(HeaderMap& headers, absl::string_view upgrade);
 
+/**
+ * The non template implementation of resolvePerFilterConfig. see
+ * resolvePerFilterConfig for docs.
+ */
+const Router::RouteSpecificFilterConfig*
+resolvePerFilterConfigGeneric(const std::string& filter_name,
+                              const Router::RouteConstSharedPtr& route);
+
+/**
+ * Retreives the route specific config. Route specific config can be in a few
+ * places, that are checked in order. The first config found is returned. The
+ * order is:
+ * - the routeEntry() (for config that's applied on weighted clusters)
+ * - the route
+ * - and finally from the virtual host object (routeEntry()->virtualhost()).
+ *
+ * To use, simply:
+ *
+ *     const auto* config =
+ *         Utility::resolvePerFilterConfig<ConcreteType>(FILTER_NAME, stream_callbacks_.route());
+ *
+ * See notes about config's lifetime below.
+ *
+ * @param filter_name The name of the filter who's route config should be
+ * fetched.
+ * @param route The route to check for route configs. nullptr routes will
+ * result in nullptr being returned.
+ *
+ * @return The route config if found. nullptr if not found. The returned
+ * pointer's lifetime is the same as the route parameter.
+ */
+template <class ConfigType>
+const ConfigType* resolvePerFilterConfig(const std::string& filter_name,
+                                         const Router::RouteConstSharedPtr& route) {
+  static_assert(std::is_base_of<Router::RouteSpecificFilterConfig, ConfigType>::value,
+                "ConfigType must be a subclass of Router::RouteSpecificFilterConfig");
+  const Router::RouteSpecificFilterConfig* generic_config =
+      resolvePerFilterConfigGeneric(filter_name, route);
+  return dynamic_cast<const ConfigType*>(generic_config);
+}
+
+/**
+ * The non template implementation of iteratePerFilterConfig. see
+ * iteratePerFilterConfig for docs.
+ */
+void iteratePerFilterConfigGeneric(
```

Comment:
This interface seems a bit more complicated than necessary. Can't we have a method to get the merged config (everything merged) and another to get the most specific configuration? Then the call site in ext_authz can just get both and use them for their respective purpose.

Reply:
Unfortunately the knowledge of how to merge is domain specific. In this method i only have access to a tagging interface which I can't do much with..

Only the caller (i.e. the filter) knows to which class to cast it do, and do the merge. My thinking here was to allow the filter iterate with guaranteed order of specificity so merge will be easy

Reply:
I'm thinking to add something like this:
```
template <class ConfigType>
absl::optional<ConfigType> getMergedPerFilterConfig(const std::string& filter_name,
                            const Router::RouteConstSharedPtr& route,
                            std::function<void(ConfigType&, const ConfigType&)> reduce) {
  absl::optional<ConfigType> merged;

  iteratePerFilterConfig<ConfigType>(
      filter_name, route, [&reduce, &merged](const ConfigType& cfg) {
        if (!merged) {
          merged.value() = cfg;
        } else {
          reduce(merged.value(), cfg);
        }
      });

  return merged;
}
```

with the reduce function merging the second arg in to the first arg so that there are less copies.

If you are ok with this, I will add tests and commit\push

Reply:
Sure, but I would then rename `iteratePer..` to just be a `fold` function, because this is logically what we're doing.

Reply:
so to summarize:
- iteratePerFilterConfig becomes foldPerFilterConfig
- add getMergedPerFilterConfig as described above
- modify the filter accordingly (I'm thinking to add the merge logic as a method to FilterConfigPerRoute)
- test everything

Reply:
Yes, I think this will be a stable and intuitive interface, thanks.

Reply:
sounds good - will add this over the weekend.
is there an appropriate place to document these methods for the next person that implements perFilterConfig?